onc-healthit/lantern-back-end
# Make endpoint manager broadcast start/stop message

Endpoint manager now broadcast the start and stop message by publishing the message to a fanout exchange (broadcast_exchange). 
In the capability querier the broadcast_queue binds to the exchange and receives the start/stop message. 
Add to .env :
`LANTERN_BROADCAST_EXCHANGE=broadcast_exchange`
`LANTERN_BROADCAST_QUEUE=broadcast_queue` 


To test:
Edit the docker-compose.yml file to include a second capabilityQuerier
copy paste the capability querier, give it a different name and pass in a different value for the `LANTERN_BROADCAST_QUEUE` env var
```  
capability_querier_2:
    build: 
      context: ./capabilityquerier
    depends_on:
      - lantern-mq
    restart: on-failure
    environment:
      - LANTERN_QUSER=${LANTERN_QUSER}
      - LANTERN_QPASSWORD=${LANTERN_QPASSWORD}
      - LANTERN_QHOST=${LANTERN_QHOST}
      - LANTERN_QPORT=${LANTERN_QPORT}
      - LANTERN_CAPQUERY_QNAME=${LANTERN_CAPQUERY_QNAME}
      - LANTERN_CAPQUERY_NUMWORKERS=${LANTERN_CAPQUERY_NUMWORKERS}
      - LANTERN_CAPQUERY_QRYINTVL=${LANTERN_CAPQUERY_QRYINTVL}
      - LANTERN_BROADCAST_EXCHANGE=${LANTERN_BROADCAST_EXCHANGE}
      - LANTERN_BROADCAST_QUEUE=broadcast_queue_2
    volumes:
      - ./scripts/wait-for-it.sh:/etc/lantern/wait-for-it.sh
    command: /etc/lantern/wait-for-it.sh lantern-mq:5672 -- ./main
```
Edit the docker-compose.override.yml file to include another capability querier
```
  capability_querier_2:
    build:
      args:
        cert_dir: ./certs
    environment:
      - LANTERN_TEST_QUSER=${LANTERN_TEST_QUSER}
      - LANTERN_TEST_QPASSWORD=${LANTERN_TEST_QPASSWORD}
      - LANTERN_TEST_QNAME=${LANTERN_TEST_QNAME}
      - LANTERN_TEST_ENDPTINFO_CAPQUERY_QNAME=${LANTERN_TEST_ENDPTINFO_CAPQUERY_QNAME}
```

`make clean`
`make run`
`make populatedb`
Go to http://localhost:15672/  and verify that under the exchanges tab you see broadcast_exchange. 
Click on broadcast_exchange and verify that you see two queues bounded to it - broadcast_queue and broadcast_queue_2 (or whatever value you defined for the second capabilityQuerier)
Send the endpoints
`go run sendendpoints/main.go`
Verify that the docker container logs shows both capability querier services making the GET request to the endpoints 

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-196
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code accomplishes the tasks purpose.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.
